Storing Prometheus data in Postgres blog

sjmiller609 opened the pull request:
We should consider if this should be launched before or after Time Series stack. This is time series data, but it's not actually using our stack.

sjmiller609 commented on src/content/blog/2024-05-20-simple-and-efficient-metrics-in-postgres/index.mdx:
```diff
@@ -0,0 +1,212 @@
+
+Leveraging **pg_partman** and **pg_cron**, we now have a simple way of efficiently storing metrics.
+
+## Tembo Time Series Stack
```
if this one goes after time series blog, this section should include a link to that blog, and perhaps a "look forward to a following blog about how we improved this system by migrating on to the time series stack"

samay-sharma replied:
Agreed. What about changing this section to something like:

"This data model helps us efficiently store time series data for this simple use case and data model. As our dataset gets larger and queries complex, we'll need more features like compression, long-term retention, and analytical functions."

At Tembo, we've built a time series stack which solves exactly this problem and helps users work with time series data more efficiently with PostgreSQL. You can store your time series data in Postgres by trying the Time Series Stack on Tembo Cloud.

We will soon be migrating this system to the time series stack to use all timeseries features Postgres and it's extensions can provide. You can look forward to our blog covering that journey in the upcoming weeks.

sjmiller609 replied:
Thank you, changed it to this

sjmiller609 commented on src/content/blog/2024-05-20-simple-and-efficient-metrics-in-postgres/index.mdx:
```diff
@@ -0,0 +1,212 @@
+
+## Prometheus Foreign Data Wrapper
+
+We also developed a Postgres extension that helps query Prometheus from inside of Postgres. Please take a look at [prometheus_fdw](https://pgt.dev/extensions/prometheus_fdw) on [GitHub](http://localhost:4321/blog/monitoring-with-prometheus-fdw), or read about it in the [blog](/blog/monitoring-with-prometheus-fdw).
```
```suggestion
We also developed a Postgres extension that helps query Prometheus from inside of Postgres. Please take a look at [prometheus_fdw](https://pgt.dev/extensions/prometheus_fdw) on [GitHub](/blog/monitoring-with-prometheus-fdw), or read about it in the [blog](/blog/monitoring-with-prometheus-fdw).
```

samay-sharma commented on src/content/blog/2024-05-20-simple-and-efficient-metrics-in-postgres/index.mdx:
```diff
@@ -0,0 +1,212 @@
+---
+slug: prometheus-data-in-postgresql
+title: How Tembo Cloud stores Prometheus data in PostgreSQL
```
Title suggestion: How Tembo Cloud stores Prometheus metrics data in PostgreSQL

samay-sharma commented on src/content/blog/2024-05-20-simple-and-efficient-metrics-in-postgres/index.mdx:
```diff
@@ -0,0 +1,212 @@
+![good-status](./good-status.png)
+![error-status](./error-status.png)
+
+This is an example of a feature that needs [time series data](https://en.wikipedia.org/wiki/Time_series).
```
I'd add a few lines here explaining what data this feature needs and why that's time series. I don't think it would be obvious to most readers.

samay-sharma commented on src/content/blog/2024-05-20-simple-and-efficient-metrics-in-postgres/index.mdx:
````diff
@@ -0,0 +1,212 @@
+
+```
+pg_isready{"instance_id"="inst_1704307947520_WRlT82_1"} 1
+pg_isready{"instance_id"="inst_1710645460091_zDBrD1_5"} 0
````
Should we obfuscate the instance id? Or make sure these are deleted instances?

samay-sharma commented on src/content/blog/2024-05-20-simple-and-efficient-metrics-in-postgres/index.mdx:
```diff
@@ -0,0 +1,212 @@
+
+This data model allows us to store all metrics data in a table **metric_values**, which we can optimize for time series data. Each row includes the timestamp, metric value, and a reference to the set of labels and metric name for that data point.
+
+In the **metric_labels** table, each row is unique by `(name, labels)`, so that we know which metric name and set of labels each data point has.
```
We should explain what a label is here. I don't think it's clear what each label is or could be.

samay-sharma commented on src/content/blog/2024-05-20-simple-and-efficient-metrics-in-postgres/index.mdx:
````diff
@@ -0,0 +1,212 @@
+]
+```
+
+## Storing metrics data in Postgres
````
I think we should mention prometheus_fdw section here. After explaining how prometheus stores data, we should explain how we import it into Postgres and could then go into the data model.

sjmiller609 replied:
The thing is, we aren't actually using prom fdw for this system

samay-sharma reviewed (commented):
Nice blog overall.

Some sections could benefit from more definitions and explanations for someone not familiar with all the concepts and I've pointed them out.

The other main feedback I had was around the section headers. We should use newspaper headline style headers as opposed to just one word because that helps in both SEO and for the reader to understand what will be covered in each section. See [Claire's slide](https://speakerdeck.com/clairegiordano/how-to-give-your-postgres-blog-posts-an-outsize-impact-claire-giordano-fosdem-2023?slide=17) for a quick checklist for improving scanability / readability of a blog.

sjmiller609 commented:
@samay-sharma thank you for your review.

I made the requested changes, and also changed the release date to 5/22/24.  I changed the base branch to ts-blog, which will automatically switch to main when ts-blog is merged.

I am happy for any copy editing changes you want to make, and you can merge this blog when you think the time is right.

samay-sharma commented:
@sjmiller609 : I did a copy-editing pass. I'll merge on Wednesday morning PST.

Please take a look Tuesday if you have any inputs, otherwise it's good to go from my side.
